components/brave_wallet/common/zcash_utils.cc

```diff
@@ -208,31 +222,117 @@ std::optional<std::string> ExtractTransparentPart(
     return std::nullopt;
   }
 
-  std::vector<uint8_t> padded_hrp(kPaddedHrpSize, 0);
-  std::copy(reinterpret_cast<const uint8_t*>(expected_hrp.c_str()),
-            reinterpret_cast<const uint8_t*>(expected_hrp.c_str()) +
-                expected_hrp.length(),
-            padded_hrp.begin());
+  auto padded_hrp = GetPaddedHRP(is_testnet);
 
   // Check that HRP is similar to the padded HRP
   if (!std::equal(padded_hrp.begin(), padded_hrp.end(),
                   reverted->end() - kPaddedHrpSize)) {
     return std::nullopt;
   }
 
-  auto parts = ParseUnifiedAddress(
+  auto parts = ParseUnifiedAddressBody(
       base::make_span(*reverted).subspan(0, reverted->size() - kPaddedHrpSize));
+
+  return parts;
+}
+
+// https://zips.z.cash/zip-0316#encoding-of-unified-addresses
+std::optional<std::string> ExtractTransparentPart(
+    const std::string& unified_address,
+    bool is_testnet) {
+  auto transparent_bytes = GetTransparentRawBytes(unified_address, is_testnet);
+  if (transparent_bytes) {
+    return PubkeyHashToTransparentAddress(transparent_bytes.value(),
+                                          is_testnet);
+  }
+  return std::nullopt;
+}
+
+std::optional<std::array<uint8_t, kOrchardRawBytesSize>> GetOrchardRawBytes(
+    const std::string& unified_address,
+    bool is_testnet) {
+  auto parts = ExtractParsedAddresses(unified_address, is_testnet);
+
   if (!parts.has_value()) {
     return std::nullopt;
   }
 
   for (const auto& part : parts.value()) {
-    if (part.first == AddrType::kP2PKH) {
-      return PubkeyHashToTransparentAddress(part.second, is_testnet);
+    if (part.first == ZCashAddrType::kOrchard) {
+      if (part.second.size() != kOrchardRawBytesSize) {
+        return std::nullopt;
+      }
+      std::array<uint8_t, kOrchardRawBytesSize> result;
+      base::ranges::copy(part.second, result.begin());
+      return result;
     }
   }
 
   return std::nullopt;
 }
 
+std::optional<std::vector<uint8_t>> GetTransparentRawBytes(
+    const std::string& unified_address,
+    bool is_testnet) {
+  auto parts = ExtractParsedAddresses(unified_address, is_testnet);
+  if (!parts.has_value()) {
+    return std::nullopt;
+  }
+
+  for (const auto& part : parts.value()) {
```

Comment:
what's the purpose in looking here if the only thing we're doing is a conditional check on the first part?

Reply:
We are choosing any transparent part from the list here